source/single_async_manual_reset_event.cpp

```diff
@@ -0,0 +1,182 @@
+  }
+
+  // resume the waiting operation
+  complete_operation(opAddr);
+}
+
+void single_async_manual_reset_event::cancel(_op_base* op) noexcept {
+  // This is only invoked from the stop_callback, so our stop_token has
+  // requested cancellation.
+  //
+  // We're about to read-modify-write state_, which we will observe to be in one
+  // of three states:
+  //
+  //  1. it was nullptr; we must be getting cancelled while we're still in
+  //     start() and start_or_wait hasn't updated state_, yet.
+  //
+  //  2. it was signalled; we could be in one of two sub-states:
+  //    a. we're still in start() and state_ was signalled before the call to
+  //       start_or_wait, or
+  //    b. start() completed by registering us to wait and cancellation lost a
+  //       race to signalling.
+  //
+  //  3. it was opAddr; we were waiting for a signal but cancellation happened
+  //     before the signal arrived.
+  //
+  // If we observe anything else, there's a bug somewhere.
+  //
+  // We're going to assume we're in state 3, and respond appropriately if not.
+  //
+  // In state 1, we need to CAS again from nullptr to cancelled so that the call
+  // to start_or_wait can observe we've been cancelled.  If the CAS from nullptr
+  // succeeds then we've been successful and we'll need to invoke set_done after
+  // destroying the callback, which we'll do in start_or_wait.  We can't enter
+  // state 3 from here because we're being cancelled before calling
+  // start_or_wait, which is the only way to update state_ to opAddr, so if the
+  // CAS from nullptr fails then we must have transitioned to state 2a.
+  //
+  // In state 2a, we'll do nothing, start_or_wait will see that state_ is
+  // signalled, and immediately invoke complete, whereupon we'll check whether
+  // the stop_token has had stop requested (which it has), and we'll invoke
+  // set_done instead of set_value.
+  //
+  // In state 2b, the signalling thread is racing with the cancelling thread,
+  // which will be synchronized in complete when we destroy the stop callback.
+  // We'll then check whether the stop token has had stop requested, find that
+  // it has, and invoke set_done so we'll do nothing here.
+  //
+  // In state 3, our first CAS will succeed, signalling in state_ that we've
+  // been cancelled.  No future signal attempt will invoke complete so we'll
+  // call unifex::set_done ourselves after destroying the callback.
+
+  const auto opAddr = to_addr(op);
+  const auto signalledState = to_addr(this);
+  const auto cancelledState = signalledState + 1;
+
+  auto expectedState = opAddr;
+
+  if (state_.compare_exchange_strong(
+      expectedState,
+      cancelledState,
+      std::memory_order_acq_rel)) {
+    // state 3
+
+    // leave it in the no-waiter state
+    clear_cancelled_state(state_, cancelledState);
+
+    complete_operation(opAddr);
+    return;
+  }
+
+  if (expectedState == to_addr(nullptr)) {
+    // state 1
+    if (state_.compare_exchange_strong(
```

Comment:
I think the order of this compare-exchange and the previous compare-exchange might need to be reversed, or at least you need to account for the possibility that this compare-exchange failed because the state transition from nullptr -> opAddr on another thread in between the previous compare-exchange and this compare-exchange and thus need to retry the opAddr -> not-signalled transition again.

The state is potentially going to transition from: not-signalled -> opAddr -> signalled concurrently with this cancel call and so you probably need to try to transition from not-signalled -> cancelled and if that fails then try opAddr -> not-signalled, and if it reaches 'signalled' then do nothing as the 'set()' call is going to be responsible for calling start().